### ISSUE TITLE
Deleting multiple schedules inside of a job template results in error

### ISSUE THREAD

**Issue by akus062381** [ansible/awx#9548]
##### ISSUE TYPE
 - Bug Report

##### SUMMARY
When you have multiple schedules added in the schedules tab of a particular job template (I added 3 for testing purposes), and you click the check boxes next to every single one, and click delete, about 50% of the time you get an error (see GIF) and the first schedule in the sequence will not delete. After you "x" the error out, you can check the box next to that one schedule and click delete and it will delete.

##### ENVIRONMENT
* AWX version: Devel 3-10-21

##### STEPS TO REPRODUCE

Click on Templates.
Click on an individual template, then click the Schedules tab.
One by one, add 3 schedules. Just a name will suffice for these purposes.
On the view of those 3 listed schedules, click the check box next to all 3 schedules. Click Delete.
About 50-60% of the time, you will get an error message pop up, and the first schedule in the list will not delete.
Note: I had to create 3 new schedules a couple of times to get the error to reproduce.

##### EXPECTED RESULTS

Checking the check boxes next to multiple schedules, and then clicking delete, would result in all of those selected schedules being deleted.

##### ACTUAL RESULTS

a 500 error was returned, and the first schedule in the list refuses to delete. Once the error is cleared, the one schedule can be successfully deleted.

##### ADDITIONAL INFORMATION

![delete error](https://user-images.githubusercontent.com/63685497/110687021-0c902d00-81ae-11eb-8b9c-7330cbda0413.gif)

**Comment by akus062381** [comment]
@jakemcdermott I was able to just recreate this error.... let me know if I can be helpful in any way.

**Comment by jakemcdermott** [comment]
@akus062381 

This is weird - I made 10 attempts to reproduce this bug ( [add 3 schedules -> delete all] x 10 ) and couldn't reproduce it.

On a whim, I switched from firefox to chrome and got it on the first try:

![Screenshot from 2021-04-20 16-37-41](https://user-images.githubusercontent.com/9753817/115460878-ca1f2e80-a1f6-11eb-95a8-8903bd799e3e.png)

**Comment by jakemcdermott** [comment]
The schedules on a JT can point to each other. If we have three schedules linked like this:

[A] -> [B] -> [C]

... and deletion requests for [A] and [B] are handled at close to the same time by different worker processes, a race condition can occur where  [A] updates with a stale reference to [B], which makes postgres sad:


![Screenshot from 2021-04-20 17-05-06](https://user-images.githubusercontent.com/9753817/115465064-05702c00-a1fc-11eb-9ee5-4c07dd35e84f.png)

**Comment by jakemcdermott** [comment]
Notes:

* The 500 means some additional exception handling is needed api-side
* Generally speaking, one should be able to delete schedules from a JT in any order they please and timing shouldn't matter.

**Comment by jakemcdermott** [comment]
* The 500 means some additional exception handling is needed
* Generally speaking, clients should be able to delete schedules in any order and timing shouldn't matter.

**Comment by jladdjr** [comment]
TLDR - think we need to synchronize schedule deletes and schedule updates. Looks like we're not doing that.

---

Looking at the error @jakemcdermott posted:

> `insert or update on table "main_unifiedjobtemplate" violates foreign key constraint "main_unifiedjobtempl_next_schedule_id_955ff55d_f..main_sche"`

> `Key (next_schedule_id)=(36) is not present in table "main_schedule"`

We were updating a unified job template
.. by setting the `next_schedule` field
.. but the schedule we were trying to set doesn't exist.

I think I see how this could play out.

The awx task scheduler updates schedules using the [awx_periodic_schedule task](https://github.com/ansible/awx/blob/devel/awx/main/tasks.py#L509)
.. which calls schedule.update_computed_fields()
.. which in turn calls a UJT's [update_computed_fields()](https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L258)

Here's a look at that method.
![image](https://user-images.githubusercontent.com/4440360/117348363-f675b480-ae5e-11eb-876b-0d6b6668faab.png)

I think this is where we set ourselves up for a race condition. The method grabs a schedule, sets `next_schedule`, and then saves. We don't take any steps to ensure that the schedule that we grabbed still exists when we save/update the UJT. I bet this is the save() that causes the PG traceback.

I think the root issue is that a schedule delete and a schedule update are not synchronized. We can issue a DELETE command and that will be taken care of at the same time that awx is busy deciding which schedule should run next. 

I'm wondering if we should wrap the [schedule's deletion](https://github.com/ansible/awx/blob/devel/awx/main/models/schedules.py#L268) in the same lock the task manager uses (`awx_periodic_scheduler_lock`) to pick the next schedule so that the two operations never overlap. So, updating this to use a lock:

![image](https://user-images.githubusercontent.com/4440360/117348933-ae0ac680-ae5f-11eb-8643-a298bb8b62ae.png)

I'm going to see if I can reproduce the issue and then try adding that lock to see if that helps clear things up.

**Comment by jladdjr** [comment]
Circling back to this. I followed up on the synchronization theory and wasn't able to fix things with locks. Will continue digging on this to see if another approach would work.

**Comment by AlanCoding** [comment]
I would suggest moving the updating of the JT's computed fields into an `on_commit` method. I say this because @jladdjr previously failure to address the race condition with obvious locks suggests to me that the problem is committing while another transaction is still open. These conflicts often can't be resolved by retrying or locking specific to the action causing the problem.

Even so, the race condition may still be possible, but in the on_commit, a simple retry or defaulting to None should work fine.
